delivery_carrier_return_barcode_pattern/models/delivery_carrier.py

```diff
@@ -0,0 +1,35 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+import re
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    return_barcode_pattern = fields.Char(help="explain here")
+
+    @api.constrains("return_barcode_pattern")
+    def _constrains_return_barcode_pattern(self):
+        for record in self:
+            if record.return_barcode_pattern:
+                try:
+                    re.compile(record.return_barcode_pattern)
+                except (re.error):
+                    raise UserError(
+                        _("Return Barcode Pattern must be a valid regular expression")
+                    )
+
+    @api.model
+    def _get_origin_from_barcode(self, barcode):
```

Comment:
don't we know the carrier when we receive the goods? Do we really have to load all carriers all the times?

Reply:
No, incoming goods might be coming from any available carrier.
I could select distinct patterns though, as I suppose most of the time it'll be empty.

Reply:
I improved this, by only using selecting distinct patterns.